pkg/webhooks/kafkacluster_validator_test.go

```diff
@@ -690,3 +693,83 @@ func TestCheckExternalListenerStartingPort(t *testing.T) {
+		kafkaClusterSpec v1beta1.KafkaClusterSpec
+		expected         field.ErrorList
+	}{
+		{
+			testName: "valid config: ingressControllerTargetPort is not defined",
+			kafkaClusterSpec: v1beta1.KafkaClusterSpec{
+				ListenersConfig: v1beta1.ListenersConfig{
+					ExternalListeners: []v1beta1.ExternalListenerConfig{
+						{
+							CommonListenerSpec: v1beta1.CommonListenerSpec{Name: "test-external1"},
+						},
+						{
+							CommonListenerSpec: v1beta1.CommonListenerSpec{Name: "test-external2"},
+						},
+					},
+				},
+			},
+			expected: nil,
+		},
+		{
+			testName: "valid config: ingressControllerTargetPort is defined and is valid for IstioIngress controller",
+			kafkaClusterSpec: v1beta1.KafkaClusterSpec{
+				ListenersConfig: v1beta1.ListenersConfig{
+					ExternalListeners: []v1beta1.ExternalListenerConfig{
+						{
+							CommonListenerSpec:          v1beta1.CommonListenerSpec{Name: "test-external1"},
+							IngressControllerTargetPort: util.Int32Pointer(int32(v1beta1.MaxWellKnownPort + rand.Intn(65536-v1beta1.MaxWellKnownPort))),
+						},
+						{
+							CommonListenerSpec:          v1beta1.CommonListenerSpec{Name: "test-external2"},
+							IngressControllerTargetPort: util.Int32Pointer(int32(v1beta1.MaxWellKnownPort + rand.Intn(65536-v1beta1.MaxWellKnownPort))),
+						},
+					},
+				},
+			},
+			expected: nil,
+		},
+		{
+			testName: "invalid config: ingressControllerTargetPort is defined and is invalid for IstioIngress controller",
+			kafkaClusterSpec: v1beta1.KafkaClusterSpec{
+				IngressController: istioingress.IngressControllerName,
+				ListenersConfig: v1beta1.ListenersConfig{
+					ExternalListeners: []v1beta1.ExternalListenerConfig{
+						{
+							CommonListenerSpec:          v1beta1.CommonListenerSpec{Name: "test-external1"},
+							IngressControllerTargetPort: util.Int32Pointer(1),
+						},
+						{
+							CommonListenerSpec:          v1beta1.CommonListenerSpec{Name: "test-external2"},
+							IngressControllerTargetPort: util.Int32Pointer(1023),
+						},
+					},
+				},
+			},
+			expected: append(field.ErrorList{},
+				field.Invalid(field.NewPath("spec").Child("listenersConfig").Child("externalListeners").Index(0).
+					Child("ingressControllerTargetPort"), int32(1),
+					invalidExternalListenerTargetPortErrMsg+": "+fmt.Sprintf(
+						"ExternalListener '%s' would use a well-known port number '%d'(<%d) as the ingress controller container port, and this is not permitted",
+						"test-external1", int32(1), v1beta1.MaxWellKnownPort)),
+				field.Invalid(field.NewPath("spec").Child("listenersConfig").Child("externalListeners").Index(1).
+					Child("ingressControllerTargetPort"), int32(1023),
+					invalidExternalListenerTargetPortErrMsg+": "+fmt.Sprintf(
+						"ExternalListener '%s' would use a well-known port number '%d'(<%d) as the ingress controller container port, and this is not permitted",
+						"test-external2", int32(1023), v1beta1.MaxWellKnownPort)),
+			),
+		},
+	}
```

Comment:
**Note**: I used to do error message unit testing exactly like this, but when the implementation changes this is going to require a similar update here - which sounds easy, but if it is not done by the same person and immediately after the logic change, can be tricky - so I switched to `errors.Errorf("")`-ing the error message I steal from the first test run output (where this fails on a dummy message) and just replace the test-case-local variables in the static string with formatting and pass them as arguments which is a bit easier to read as non-author and also to update. I can even make a case for using the static string, but that's up to personal preference, I wouldn't comment that, such a difference is easy to track down inside the test case.

I'm not advocating for or against any of these approaches, just wanted to share my experience.

Reply:
Thx, I also found that the current way of matching the test outputs to the actual outputs is very cumbersome, will try to use the easy way tmr

Reply:
hey @pregnor - I tried with something `append(field.ErrorList{}, errors.Errorf("test"))`, but realized `field.ErrorList{}` is expecting `[]*Error` (with the its own [`Error` type](https://pkg.go.dev/k8s.io/apimachinery@v0.26.4/pkg/util/validation/field#Error)), and the type seems to be incompatible with `errors.Errorf("")`

Reply:
Sad, thanks for trying.